SudokuSolver/SudokuSolver/SudokuSolver.cs

```diff
@@ -0,0 +1,137 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Toolkit.Extensions;
+
+namespace SudokuSolver
+{
+    public class SudokuSolver
+    {
+        public class EmptySudokuCube
+        {
+            public int Row { get; init; }
+            public int Col { get; init; }
+
+            public List<int> PossibleValues { get; init; }
+            public EmptySudokuCube(int row, int col)
+            {
+                Row = row;
+                Col = col;
+                PossibleValues = new List<int>();
+            }
+        }
+
+        private int[,] _grid;
+        private int _maxRows;
+        private int _maxCols;
+        private int _rowSqrt;
+        private int _colSqrt;
+
+        private List<EmptySudokuCube> _emptyCubes;
+
+        private void InitEmptyCubes()
+        {
+            _emptyCubes = new List<EmptySudokuCube>();
+
+            for (int i = 0; i < _maxRows; i++)
+            {
+                for (int j = 0; j < _maxCols; j++)
+                {
+                    if (_grid[i, j] == 0)
+                    {
+                        var eSC = new EmptySudokuCube(i, j);
+                        // Check all legal nums and add them to cube's possible values
+                        eSC.PossibleValues.AddRange(
+                            Enumerable.Range(1, _maxRows).Where(num => IsLegal(i, j, num)).ToList()
+                        );
+                        _emptyCubes.Add(eSC);
+                    }
+                }
+            }
+        }
+
+        public SudokuSolver(int [, ] grid)
+        {
+            _grid = grid;
+            _maxRows = grid.GetLength(0);
+            _maxCols = grid.GetLength(1);
+
+            if (_maxCols != _maxRows) throw new ArgumentException("Matrix should have the same number of rows and columns.");
```

Comment:
Tutaj jest to za mały warunek. Algorytm nie zadziała dla gridów, których rozmiar nie jest kwadratem liczby naturalnej